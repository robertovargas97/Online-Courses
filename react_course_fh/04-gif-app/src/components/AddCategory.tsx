import React from "react";
import { useState } from "react";

interface addCategoryProps {
    gifsCategories: string[],
    setGifsCategories: React.Dispatch<React.SetStateAction<string[]>>
}


export const AddCategory: React.FC<addCategoryProps> = ({ gifsCategories, setGifsCategories }): JSX.Element => {

    type OnChangeFunctionType = (event: React.ChangeEvent<HTMLInputElement>) => void;
    type OnSubmitFunctionType = (event: React.FormEvent<HTMLFormElement>) => void;
    type OnAddCategoryType = (newCategory: string) => void;
    type ValidateInputType = (sanitizedInputValue: string | undefined) => boolean;

    const [inputValue, setInputValue] = useState("");

    const validateInput: ValidateInputType = (sanitizedInputValue) => {
        let isValid = false;

        if (sanitizedInputValue) {

            if (sanitizedInputValue.length > 1) {
                if (!(gifsCategories.find(category => inputValue.toUpperCase() === category.toUpperCase()))) {
                    isValid = false;
                }
                isValid = true;
            }
        }

        return isValid;
    }

    const addNewCategory: OnAddCategoryType = (newCategory) => {
        // ... is used to extract/copy the array
        const newArray: string[] = [newCategory, ...gifsCategories];
        setGifsCategories(newArray);
    }


    const onInputChange: OnChangeFunctionType = (event): void => {
        const data: string = event.target.value;
        setInputValue(data);
    }


    const onSubmit: OnSubmitFunctionType = (event) => {
        event.preventDefault();
        const sanitizedInputValue = inputValue.trim();

        const isValidCategory = validateInput(sanitizedInputValue);
        if (isValidCategory) {
            addNewCategory(sanitizedInputValue);
            setInputValue("");
        }
        else {
            //TODO: Change for something more elegant
            console.error('There was an issue');  
        }
    }


    const addCategoryComponent: JSX.Element =

        <form data-testid="form" action="" method="get" onSubmit={onSubmit}>
            <input data-testid="form-input" type="text" placeholder="Search Gifs" value={inputValue} onChange={onInputChange} />
        </form>


    return addCategoryComponent;
}
