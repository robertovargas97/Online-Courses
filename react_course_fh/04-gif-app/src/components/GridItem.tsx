import React from 'react'

interface GridItemProps {
    title: string,
    url: string
}

export const GridItem: React.FC<GridItemProps> = ({ title, url }): JSX.Element => {

    const gridItemComponent =
        <div className='card'>
            <img data-testid="gif" src={url} alt={title} />
            <p data-testid="image-title">{title}</p>
        </div>;
    return gridItemComponent;
}
