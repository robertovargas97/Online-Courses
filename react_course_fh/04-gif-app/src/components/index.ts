export * from './AddCategory';
export * from './GifsGrid';
export * from './GridItem';