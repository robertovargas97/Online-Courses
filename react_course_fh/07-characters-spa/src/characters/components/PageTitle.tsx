import React from 'react'

type PageTitleProps = {
  title: string
}
export const PageTitle: React.FC<PageTitleProps> = ({ title }) => {
  const component = (
    <>
      <h1 className="mt-4 text-center"> {title} </h1>
      <hr />
    </>
  )
  return component
}
