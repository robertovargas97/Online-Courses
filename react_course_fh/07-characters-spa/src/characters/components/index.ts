export * from './CharactersList'
export * from './CharacterCard'
export * from './PageTitle'
