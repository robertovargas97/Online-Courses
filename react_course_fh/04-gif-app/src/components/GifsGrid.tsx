import { GridItem } from "./GridItem";
import { useFetchGifs } from "../hooks/useFetchGifs";


interface GifsGridProps {
    category: string
}

export const GifsGrid: React.FC<GifsGridProps> = ({ category }): JSX.Element => {


    const { gifs, isLoading } = useFetchGifs(category);

    const gridComponent =
        <>
            <h3>
                {category}
            </h3>
            {
                isLoading && (<h2 data-testid="loading"> Loading...</h2>)
            }


            <div className="card-grid">
                {
                    gifs.map(
                        (gif) => (
                            <GridItem key={gif.id} {...gif} />
                        )
                    )
                }
            </div>
        </>;
    return gridComponent;
}


