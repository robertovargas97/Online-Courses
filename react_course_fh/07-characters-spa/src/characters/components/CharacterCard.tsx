import React from 'react'
import { Character } from '../data/characters'

import './card.styles.css'
import { Link } from 'react-router-dom'

type CharacterCardProps = {
  character: Character
  animeLogo: string
}

export const CharacterCard: React.FC<CharacterCardProps> = ({
  character,
  animeLogo,
}) => {
  const characterImgUrl = `/assets/${character.id}.jpg`
  const component = (
    <>
      <div className="character">
        <div className="main">
          <img
            className="characterImage"
            src={characterImgUrl}
            alt="character"
          />
          <h2 className="mt-2">{character.name} </h2>
          <p className="first-appereance">First Appereance:</p>{' '}
          {character.first_appearance}
          <Link to={`/character/${character.id}`}>More...</Link>
          <hr className="mt-3" />
          <div className="anime">
            <div className="wrapper">
              <img src={animeLogo} alt="anime" />
            </div>
            <p>
              <ins>Anime: </ins> {character.anime}
            </p>
          </div>
        </div>
      </div>
    </>
  )
  return component
}
