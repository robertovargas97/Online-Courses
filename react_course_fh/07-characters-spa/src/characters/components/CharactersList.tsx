import React, { useMemo } from 'react'
import { getCharactersByAnime } from '../helpers'
import { CharacterCard } from './CharacterCard'

type CharactersListPropts = {
  anime: string
  logoUrl: string
}

export const CharactersList: React.FC<CharactersListPropts> = ({
  anime,
  logoUrl,
}) => {
  const characters = useMemo(() => getCharactersByAnime(anime), [anime])

  const component = (
    <div className="row rows-cols-1 row-cols-md-3 g-3 d-flex justify-content-center">
      {characters.map((character) => (
        <CharacterCard
          key={character.id}
          character={character}
          animeLogo={logoUrl}
        />
      ))}
    </div>
  )

  return component
}
