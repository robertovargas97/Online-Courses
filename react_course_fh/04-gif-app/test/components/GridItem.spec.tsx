import { GridItem } from '../../src/components/GridItem';
import { render, screen } from '@testing-library/react';



describe('<GridItem> Component Test Suite', () => {
    const title: string = "Saitma";
    const url: string = "https://one-punch.com/";


    test('should match the snapshoop', () => {
        const { container } = render(<GridItem title={title} url={url} />);
        expect(container).toMatchSnapshot();
    });

    test('should display the image with the given url and title', () => {
        render(<GridItem title={title} url={url} />);
        // See the component to test
        // screen.debug();
        const gifElement: HTMLImageElement = screen.getByTestId('gif');
        const { src: gifElementUrl, alt: gifElementAltText } = gifElement;

        expect(gifElementUrl).toBe(url);
        expect(gifElementAltText).toBe(title);

    });

    test('should display the image tittle as <p>', () => {
        render(<GridItem title={title} url={url} />);

        const imageTitle: string = screen.getByTestId("image-title").innerHTML;
        expect(imageTitle).toBe(title);

    });







});