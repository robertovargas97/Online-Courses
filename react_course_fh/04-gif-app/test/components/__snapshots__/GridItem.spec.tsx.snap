// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GridItem> Component Test Suite should match the snapshoop 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="Saitma"
      data-testid="gif"
      src="https://one-punch.com/"
    />
    <p
      data-testid="image-title"
    >
      Saitma
    </p>
  </div>
</div>
`;
